### ISSUE TITLE
efficientnet-b8 and AdvProp

### ISSUE THREAD

**Issue by seefun** [lukemelas/EfficientNet-PyTorch#115]
With advprop, efficientnet got greater score in ImageNet. Would you update to the new ckpt?
the paper: https://arxiv.org/pdf/1911.09665.pdf
https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet

**Comment by lukemelas** [comment]
Awesome, thanks for posting this. It's on the way.

**Comment by Oktai15** [comment]
@lukemelas maybe you current code in “tf_to_pytorch” dir will work?

**Comment by laurentdillard** [comment]
@lukemelas  any update on that ?

**Comment by jionie** [comment]
@lukemelas any update on that ?

**Comment by seefun** [comment]
![1576659729](https://user-images.githubusercontent.com/19152032/71071901-6cb93e80-21b8-11ea-9a08-957dc00a266a.jpg)

**Comment by lukemelas** [comment]
Apologies for the delay on this (I had final exams this past week). Coming soon.

**Comment by laurentdillard** [comment]
Hi @lukemelas , just wondering if you had the chance to work on that one.

**Comment by lukemelas** [comment]
Sorry this took forever. It should be in now :) 

Let me know if you have any issues.

**Comment by lukemelas** [comment]
Closing this, but feel free to re-open it if you have any issues/questions.

**Comment by sliawatimena** [comment]
Dear all,

I am using https://colab.research.google.com/drive/1Jw28xZ1NJq4Cja4jLe6tJ6_F5lCzElb4
Why the efficientnet-b0 advprop=True result is very bad?
Thank you.

Suryadi

```
Loaded pretrained weights for efficientnet-b0, advprop=False
-----
giant panda, panda, panda bear, coon bear, Ailuropoda melanoleuca           (83.44%)
brown bear, bruin, Ursus arctos                                             (0.62%)
lesser panda, red panda, panda, bear cat, cat bear, Ailurus fulgens         (0.60%)
ice bear, polar bear, Ursus Maritimus, Thalarctos maritimus                 (0.44%)
Arctic fox, white fox, Alopex lagopus                                       (0.34%)

Loaded pretrained weights for efficientnet-b0, advprop=True
-----
wombat                                                                      (3.34%)
candle, taper, wax light                                                    (2.00%)
Angora, Angora rabbit                                                       (1.87%)
schipperke                                                                  (1.86%)
hog, pig, grunter, squealer, Sus scrofa                                     (1.60%)
```

**Comment by lukemelas** [comment]
Did you use the advprop image preprocessing or the usual preprocessing? See https://github.com/lukemelas/EfficientNet-PyTorch/blob/master/examples/imagenet/main.py#L211. That's the reason advprop is not enabled by default. Let me know if it still doesn't work and I can look into it.

**Comment by shijianjian** [comment]
> Did you use the advprop image preprocessing or the usual preprocessing? See https://github.com/lukemelas/EfficientNet-PyTorch/blob/master/examples/imagenet/main.py#L211. That's the reason advprop is not enabled by default. Let me know if it still doesn't work and I can look into it.

Hi @lukemelas ,

Thanks for the repo. Do you know the reason for using a different normalization for advprop? If I am training a new model with advprop, why should I use it than imagenet mean and std?

**Comment by ooodragon94** [comment]
one question, if I'm getting the gist of the paper right, It seems like advprop uses two batchnorm layer (one for standard data and the other one for adversarial data). However in the code, I don't see where it implements that _other batchnorm layer_. Am I misunderstanding the paper? Or is the code not providing it?

**Comment by shijianjian** [comment]
@ooodragon94 I think that part is not in this repo. But it is easy to implement. 
```python
Class EfficientNet():
    def __init__(self, advprop=False, **kwargs):
        self.somelayers = nn.Layer
        self.norm = nn.BatchNorm
        if advprop:
            self.aux_norm = nn.BatchNorm

    def forward(self, x, advprop=False):
        x = self.somelayers(x)
        if advprop:
            x = self.aux_norm(x)
        else:
            x = self.norm(x)
```

**Comment by ooodragon94** [comment]
@shijianjian 
you are definitely right (and your code is compact and beautiful!)
just that if I were to use pretrained model and fine tune it, I'm assuming that since I am not loading aux_norm's parameters, the performance might degrade due to different settings.

**Comment by shijianjian** [comment]
@ooodragon94 
Not pretty confident. But if you want to fine-tune with adversarial examples. I think the easiest way is to freeze the whole model apart from the auxiliary normalization layer. In initial epochs, you only train the auxiliary mean and stds. Then you may save the whole model and fine-tune it as normal.

**Comment by ooodragon94** [comment]
@shijianjian 
I'm back to this advprop
Seems like the paper uses both clean and adversary samples
not sure why this repo only uses adversary samples to train efficientnet
I totally agree with your comment on "different normalization"

**Comment by feiwofeifeixiaowo** [comment]
When using advprop pretrained weight and advprop normalization, the training results become very unstable.

**Comment by ooodragon94** [comment]
> When using advprop pretrained weight and advprop normalization, the training results become very unstable, and the accuracy also decreases.

@feiwofeifeixiaowo If you didn't by any change implemented advprop but only loaded its weights, than it will definitely suffer in training accuracy.
